### PR TITLE
cctools 895

### PR DESCRIPTION
**Pull request by al3xtjames**
- [x] Have you followed the [guidelines for contributing](https://github.com/Homebrew/homebrew-core/blob/master/CONTRIBUTING.md)?
- [x] Have you checked that there aren't other open [pull requests](https://github.com/Homebrew/homebrew-core/pulls) for the same formula update/change?
- [x] Have you built your formula locally with `brew install --build-from-source <formula>`, where `<formula>` is the name of the formula you're submitting?
- [x] Does your build pass `brew audit --strict <formula>` (after doing `brew install <formula>`)?

-----

This PR depends on https://github.com/Homebrew/formula-patches/pull/239; the placeholder patches (gists) will be replaced once it is merged.

### CODE REVIEW

**Comment by mistydemeo** [comment]
@BrewTestBot test this please

**Review by mistydemeo — APPROVED** [review]

**Comment by mistydemeo** [comment]
Hm, looks like the test is failing due to a missing executable:

```
        Testing cctools
/usr/bin/sandbox-exec -f /tmp/homebrew20180514-94286-1s1tnzi.sb /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3_2/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/test.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/cctools.rb --verbose
==> /usr/local/Cellar/cctools/895/bin/otool -L /usr/local/Cellar/cctools/895/bin/install_name_tool
fatal error: /usr/local/Cellar/cctools/895/bin/otool: can't find or exec: /usr/local/Cellar/cctools/895/bin/llvm-objdump (No such file or directory)
fatal error: /usr/local/Cellar/cctools/895/bin/otool: internal objdump command failed
Error: cctools: failed
<0> expected but was
<1>.
```

**Review by mistydemeo — CHANGES_REQUESTED** [review]
Hm, looks like the test is failing due to a missing executable:

```
        Testing cctools
/usr/bin/sandbox-exec -f /tmp/homebrew20180514-94286-1s1tnzi.sb /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3_2/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/test.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/cctools.rb --verbose
==> /usr/local/Cellar/cctools/895/bin/otool -L /usr/local/Cellar/cctools/895/bin/install_name_tool
fatal error: /usr/local/Cellar/cctools/895/bin/otool: can't find or exec: /usr/local/Cellar/cctools/895/bin/llvm-objdump (No such file or directory)
fatal error: /usr/local/Cellar/cctools/895/bin/otool: internal objdump command failed
Error: cctools: failed
<0> expected but was
<1>.
```

**Comment by al3xtjames** [comment]
There appears to be two issues here:
 - cctools-895 always has `-DLLVM_OTOOL` defined, so it will always assume otool from LLVM is available (shoud only be done if `--with-llvm` is specified)
 - cctools-895 tries to launch `llvm-objdump` from its current path; this should probably be patched the Homebrew prefix

I'll work on patching this soon.

**Comment by al3xtjames** [comment]
Apologies for the delay. Updated the commit, cctools' version of `otool` should now function (with and without LLVM).

**Comment by mistydemeo** [comment]
Thanks! Looks like it worked on 10.12 and 10.13, but I'm seeing a 10.11 failure:

```
        Testing cctools
/usr/bin/sandbox-exec -f /tmp/homebrew20180604-99902-1lew18o.sb /usr/local/Homebrew/Library/Homebrew/vendor/portable-ruby/2.3.3_2/bin/ruby -W0 -I /usr/local/Homebrew/Library/Homebrew -- /usr/local/Homebrew/Library/Homebrew/test.rb /usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/cctools.rb --verbose
==> /usr/local/Cellar/cctools/895/bin/otool -L /usr/local/Cellar/cctools/895/bin/install_name_tool
fatal error: /usr/local/Cellar/cctools/895/bin/otool: can't find or exec: @@LLVM_BINDIR@@/llvm-objdump (No such file or directory)
fatal error: /usr/local/Cellar/cctools/895/bin/otool: internal objdump command failed
Error: cctools: failed
<0> expected but was
<1>.
```

**Comment by al3xtjames** [comment]
Updated the `objdump` path (`/Library/Developer/CommandLineTools/usr/bin/objdump`) to fix the 10.11 failure (still works on 10.12/10.13). Is it OK to hardcode the CommandLineTools path?

File: Formula/cctools.rb [hunk 1]
```diff
@@ -47,10 +47,16 @@ class Cctools < Formula
...
-      sha256 "3b687f2b9388ac6c4acac2b7ba28d9fd07f2a16e7d2dad09aa2255d98ec1632b"
+    # Fix building strings with LTO disabled
+    patch :p0 do
+      url "https://gist.githubusercontent.com/al3xtjames/8d68ef87c4c2138ef89d3d6e20879401/raw/d178e955c0f6cfcb91d775c016cb5ad95da43db0/cctools-895-strings-no-lto.diff"
+      sha256 "885302f6573d7bf472f7bff4600b6387234df9aea8cd4187e736e0fe89551eab"
+    end
+
+    # Fix objdump/llvm-objdump paths in otool
+    patch :p0 do
+      url "https://gist.githubusercontent.com/al3xtjames/ac4096fa2508af7024ba410fef67b9df/raw/5d601f9508e39249220ab7cd2b4fb310087c4969/cctools-895-otool-objdump-path.diff"
```
Comment (mistydemeo) [inline Formula/cctools.rb]: How about setting this to a placeholder string, like `@@CLT_BINDIR@@`, that you can then `inreplace` within the formula? e.g.

```ruby
inreplace "otool/main.c", "@@CLT_BINDIR@@", "#{MacOS.active_developer_dir}/usr/bin"
```

**Review by mistydemeo — CHANGES_REQUESTED** [review]

**Comment by mistydemeo** [comment]
I guess those tools aren't actually shipped in Xcode's SDKs. Sorry for the bad suggestion. :/ Maybe just `/usr/bin/objdump`?

**Comment by al3xtjames** [comment]
IIRC `/usr/bin/objdump` was present in 10.12 and 10.13, but not 10.11, while `/Library/Developer/CommandLineTools/usr/bin/objdump` is present on all three.

**Comment by al3xtjames** [comment]
Kept the` CLT_BINDIR` patch in the latest patchset, but changed the formula to replace it with `/Library/Developer/CommandLineTools/usr/bin/objdump`.

**Comment by al3xtjames** [comment]
Added some more patching to fix cctools' version of `as` (needs a fixed path to `clang`).

**Review by mistydemeo — COMMENTED** [review]
Thanks so much for making that change!

Right now this has introduced a CLT dependency. That's maybe an acceptable compromise, but should get marked as `depends_on :clt` if so. Maybe we can add that dependency for 10.7 through 10.11, and depend on the versions in `/usr/bin` or `"#{MacOS::Xcode.prefix}/Toolchains/XcodeDefault.xctoolchain/usr/bin"` for 10.12+?

Sorry to add complexity!

**Comment by ilovezfs** [comment]
```
depends_on :clt
```
Does that exist?

**Comment by al3xtjames** [comment]
`depends_on :clt`

It looks like this was removed recently (in https://github.com/Homebrew/brew/commit/cffa5a9864e771f19231537b812b9b32a1160be0):

```
Error: Calling 'depends_on :clt' is disabled!
There is no replacement.
/usr/local/Homebrew/Library/Taps/homebrew/homebrew-core/Formula/cctools.rb:33:in `<class:Cctools>'
```

**Comment by mistydemeo** [comment]
@ilovezfs Custom requirement for this one formula, maybe? Given it's a part of Apple's developer tools, it makes sense to me that this one formula will have weird angles that we don't otherwise want.

**Comment by ilovezfs** [comment]
Is it also possible to build with Xcode only? If so, I think we should use `pour_bottle? do` as in the gcc formula, and in the source build case make the Xcode-only tweaks if without CLT.

**Pull request closed without merging**